File: src/lib/vector/vector-client.ts
import { Pinecone } from '@pinecone-database/pinecone';
import { AI_CONFIG, getPineconeApiKey } from '@/lib/ai/config';
import { getLLMClient } from '@/lib/ai/llm-client';
import { getCompanies, searchPhonesByModel, getAllPhones, getPhoneById, getFilteredPhones } from '@/lib/db/database';
import { VectorSearchResult, Phone, Filters } from '@/types';

/**
 * Vector Client for Pinecone
 * Used for typo correction and semantic similarity
 * Ported from Python implementation
 */
export class VectorClient {
  private pinecone: Pinecone;
  private indexName: string;
  private initialized: boolean = false;
  private hasApiKey: boolean = false;

  constructor() {
    const apiKey = getPineconeApiKey();
    this.hasApiKey = !!apiKey;

    if (this.hasApiKey) {
      this.pinecone = new Pinecone({ apiKey });
    } else {
      // Create a dummy instance - will use fallback mode
      this.pinecone = null as unknown as Pinecone;
    }
    this.indexName = AI_CONFIG.PINECONE_INDEX_NAME;
  }

  /**
   * Initialize the Pinecone index
   */
  async initIndex(): Promise<void> {
    if (this.initialized) return;

    // Skip if no API key - use fallback mode
    if (!this.hasApiKey) {
      this.initialized = false;
      return;
    }

    try {
      // Check if index exists
      const indexes = await this.pinecone.listIndexes();
      const indexExists = indexes.indexes?.some(idx => idx.name === this.indexName);

      if (!indexExists) {
        // Create index
        await this.pinecone.createIndex({
          name: this.indexName,
          dimension: AI_CONFIG.EMBEDDING_DIM,
          metric: 'cosine',
          spec: {
            serverless: {
              cloud: 'aws',
              region: 'us-east-1'
            }
          }
        });

        // Wait for index to be ready
        await this.waitForIndex();
      }

      this.initialized = true;
    } catch (error) {
      console.error('Failed to initialize Pinecone index:', error);
      // Continue without vector search - graceful degradation
      this.initialized = false;
    }
  }

  /**
   * Wait for index to be ready
   */
  private async waitForIndex(): Promise<void> {
    const maxAttempts = 30;
    for (let i = 0; i < maxAttempts; i++) {
      const description = await this.pinecone.describeIndex(this.indexName);
      if (description.status?.ready) {
        return;
      }
      await new Promise(resolve => setTimeout(resolve, 2000));
    }
    throw new Error('Timeout waiting for Pinecone index');
  }

  /**
   * Find similar items for typo correction
   */
  async findSimilar(
    query: string,
    typeFilter?: 'company' | 'model',
    companyFilter?: string,
    threshold: number = 0.7
  ): Promise<VectorSearchResult[]> {
    if (!this.initialized) {
      await this.initIndex();
    }

    if (!this.initialized) {
      // Fallback to fuzzy matching without vectors
      return this.fallbackSimilar(query, typeFilter);
    }

    try {
      const llm = getLLMClient();
      const embedding = await llm.embed(query.toLowerCase());

      const index = this.pinecone.index(this.indexName);

      // Build filter
      const filter: Record<string, unknown> = {};
      if (typeFilter) {
        filter.type = typeFilter;
      }
      if (companyFilter) {
        filter.company = companyFilter.toLowerCase();
      }

      const results = await index.query({
        vector: embedding,
        topK: 5,
        filter: Object.keys(filter).length > 0 ? filter : undefined,
        includeMetadata: true
      });

      return (results.matches || [])
        .filter(match => match.score && match.score >= threshold)
        .map(match => ({
          id: match.id,
          score: match.score || 0,
          metadata: match.metadata as VectorSearchResult['metadata']
        }));
    } catch (error) {
      console.error('Vector search failed:', error);
      return this.fallbackSimilar(query, typeFilter);
    }
  }

  /**
   * Fallback similarity using simple string matching
   */
  private async fallbackSimilar(
    query: string,
    typeFilter?: 'company' | 'model'
  ): Promise<VectorSearchResult[]> {
    const queryLower = query.toLowerCase();
    const results: VectorSearchResult[] = [];

    if (!typeFilter || typeFilter === 'company') {
      const companies = await getCompanies();
      for (const company of companies) {
        const score = this.stringSimilarity(queryLower, company.toLowerCase());
        if (score > 0.5) {
          results.push({
            id: `company-${company}`,
            score,
            metadata: { type: 'company', value: company }
          });
        }
      }
    }

    if (!typeFilter || typeFilter === 'model') {
      const phones = await searchPhonesByModel(query);
      for (const phone of phones.slice(0, 5)) {
        const score = this.stringSimilarity(queryLower, phone.model_name.toLowerCase());
        results.push({
          id: `model-${phone.id}`,
          score: Math.max(score, 0.6),
          metadata: {
            type: 'model',
            value: phone.model_name,
            company: phone.company_name
          }
        });
      }
    }

    return results.sort((a, b) => b.score - a.score).slice(0, 5);
  }

  /**
   * Simple string similarity (Levenshtein-based)
   */
  private stringSimilarity(s1: string, s2: string): number {
    const longer = s1.length > s2.length ? s1 : s2;
    const shorter = s1.length > s2.length ? s2 : s1;

    if (longer.length === 0) return 1.0;

    const distance = this.levenshteinDistance(longer, shorter);
    return (longer.length - distance) / longer.length;
  }

  /**
   * Levenshtein distance
   */
  private levenshteinDistance(s1: string, s2: string): number {
    const costs: number[] = [];
    for (let i = 0; i <= s1.length; i++) {
      let lastValue = i;
      for (let j = 0; j <= s2.length; j++) {
        if (i === 0) {
          costs[j] = j;
        } else if (j > 0) {
          let newValue = costs[j - 1];
          if (s1.charAt(i - 1) !== s2.charAt(j - 1)) {
            newValue = Math.min(Math.min(newValue, lastValue), costs[j]) + 1;
          }
          costs[j - 1] = lastValue;
          lastValue = newValue;
        }
      }
      if (i > 0) costs[s2.length] = lastValue;
    }
    return costs[s2.length];
  }

  /**
   * Build index with company and model data (legacy - for typo correction)
   */
  async buildIndex(): Promise<void> {
    if (!this.initialized) {
      await this.initIndex();
    }

    if (!this.initialized) {
      return;
    }

    try {
      const llm = getLLMClient();
      const index = this.pinecone.index(this.indexName);

      // Index companies
      const companies = await getCompanies();
      const companyVectors = [];

      for (const company of companies) {
        const embedding = await llm.embed(company);
        companyVectors.push({
          id: `company-${company}`,
          values: embedding,
          metadata: { type: 'company', value: company }
        });
      }

      if (companyVectors.length > 0) {
        await index.upsert(companyVectors);
      }
    } catch (error) {
      console.error('Failed to build vector index:', error);
    }
  }

  /**
   * Build product index with rich phone descriptions for RAG
   * This embeds all phones with their full specs for semantic search
   */
  async buildProductIndex(onProgress?: (current: number, total: number) => void): Promise<{ success: boolean; count: number; message: string }> {
    if (!this.hasApiKey) {
      return { success: false, count: 0, message: 'Pinecone API key not configured' };
    }

    if (!this.initialized) {
      await this.initIndex();
    }

    if (!this.initialized) {
      return { success: false, count: 0, message: 'Failed to initialize Pinecone index' };
    }

    try {
      const llm = getLLMClient();
      const index = this.pinecone.index(this.indexName);
      const phones = await getAllPhones();

      // Process in batches of 10 to avoid rate limits
      const batchSize = 10;
      let indexed = 0;

      for (let i = 0; i < phones.length; i += batchSize) {
        const batch = phones.slice(i, i + batchSize);
        const vectors = [];

        for (const phone of batch) {
          // Create rich text description for embedding
          const description = this.createPhoneDescription(phone);
          const embedding = await llm.embed(description);

          if (embedding.length > 0) {
            vectors.push({
              id: `product-${phone.id}`,
              values: embedding,
              metadata: {
                type: 'product',
                id: phone.id,
                company: phone.company_name,
                model: phone.model_name,
                price: phone.price_inr,
                ram: phone.ram_gb,
                battery: phone.battery_mah,
                camera: phone.back_camera_mp,
                rating: phone.user_rating,
                description: description.substring(0, 500) // Truncate for metadata
              }
            });
          }
        }

        if (vectors.length > 0) {
          await index.upsert(vectors);
          indexed += vectors.length;
        }

        onProgress?.(indexed, phones.length);

        // Small delay between batches to respect rate limits
        await new Promise(resolve => setTimeout(resolve, 200));
      }

      return { success: true, count: indexed, message: `Indexed ${indexed} phones successfully` };
    } catch (error) {
      console.error('Failed to build product index:', error);
      return { success: false, count: 0, message: error instanceof Error ? error.message : 'Unknown error' };
    }
  }

  /**
   * Create rich text description for a phone for embedding
   */
  private createPhoneDescription(phone: Phone): string {
    const features: string[] = [];

    // Camera features
    if (phone.back_camera_mp >= 100) features.push('flagship camera', '100MP+', 'pro photography');
    else if (phone.back_camera_mp >= 64) features.push('high resolution camera', '64MP+', 'good photography');
    else if (phone.back_camera_mp >= 48) features.push('quality camera', '48MP');

    if (phone.front_camera_mp >= 32) features.push('great selfie camera', 'vlogging', 'video calling');
    else if (phone.front_camera_mp >= 16) features.push('good front camera', 'selfies');

    // Gaming/Performance features
    if (phone.ram_gb >= 12) features.push('gaming powerhouse', 'multitasking', 'heavy gaming', 'BGMI', 'Free Fire');
    else if (phone.ram_gb >= 8) features.push('smooth gaming', 'good performance');
    else if (phone.ram_gb >= 6) features.push('casual gaming', 'everyday use');

    // Battery features
    if (phone.battery_mah >= 6000) features.push('massive battery', 'two-day battery', 'power user');
    else if (phone.battery_mah >= 5000) features.push('all-day battery', 'long lasting');
    else if (phone.battery_mah >= 4500) features.push('good battery life');

    // Storage features
    if (phone.memory_gb >= 256) features.push('lots of storage', 'store many apps', 'photos and videos');
    else if (phone.memory_gb >= 128) features.push('good storage', '128GB');

    // Price category features
    if (phone.price_inr < 15000) features.push('budget friendly', 'affordable', 'value for money', 'entry level');
    else if (phone.price_inr < 25000) features.push('mid-range', 'balanced', 'good value');
    else if (phone.price_inr < 40000) features.push('upper mid-range', 'premium features');
    else if (phone.price_inr < 60000) features.push('flagship', 'premium', 'high-end');
    else features.push('ultra premium', 'flagship killer', 'top of the line');

    // Screen features
    if (phone.screen_size >= 6.7) features.push('large display', 'big screen', 'media consumption', 'presentations');
    else if (phone.screen_size >= 6.4) features.push('comfortable display');
    else if (phone.screen_size <= 6.0) features.push('compact', 'one-hand use');

    // Rating features
    if (phone.user_rating >= 4.5) features.push('highly rated', 'user favorite', 'top rated');
    else if (phone.user_rating >= 4.0) features.push('well reviewed', 'good ratings');

    return `${phone.company_name} ${phone.model_name}. ` +
      `Price ₹${phone.price_inr.toLocaleString('en-IN')}. ` +
      `${phone.ram_gb}GB RAM, ${phone.memory_gb}GB storage. ` +
      `${phone.back_camera_mp}MP rear camera, ${phone.front_camera_mp}MP front camera. ` +
      `${phone.battery_mah}mAh battery. ` +
      `${phone.screen_size}" display. ` +
      `Processor: ${phone.processor}. ` +
      `Rating: ${phone.user_rating}/5. ` +
      `Features: ${features.join(', ')}.`;
  }

  /**
   * Semantic search for products using natural language query
   * This is the core RAG retrieval function
   */
  async searchProducts(
    query: string,
    filters?: {
      maxPrice?: number;
      minPrice?: number;
      minRam?: number;
      minBattery?: number;
      company?: string;
    },
    topK: number = 5
  ): Promise<Phone[]> {
    // If no API key or not initialized, fall back to database search
    if (!this.hasApiKey || !this.initialized) {
      return this.fallbackProductSearch(query, filters, topK);
    }

    try {
      const llm = getLLMClient();
      const embedding = await llm.embed(query);

      if (embedding.length === 0) {
        return this.fallbackProductSearch(query, filters, topK);
      }

      const index = this.pinecone.index(this.indexName);

      // Build Pinecone filter
      const pineconeFilter: Record<string, unknown> = { type: 'product' };

      if (filters?.maxPrice) {
        pineconeFilter.price = { $lte: filters.maxPrice };
      }
      if (filters?.minPrice) {
        pineconeFilter.price = { ...(pineconeFilter.price as object || {}), $gte: filters.minPrice };
      }
      if (filters?.minRam) {
        pineconeFilter.ram = { $gte: filters.minRam };
      }
      if (filters?.minBattery) {
        pineconeFilter.battery = { $gte: filters.minBattery };
      }
      if (filters?.company) {
        pineconeFilter.company = filters.company.toLowerCase();
      }

      const results = await index.query({
        vector: embedding,
        topK: topK,
        filter: pineconeFilter,
        includeMetadata: true
      });

      // Fetch full phone data from database
      const phones: Phone[] = [];
      for (const match of results.matches || []) {
        const phoneId = match.metadata?.id as number;
        if (phoneId) {
          const phone = await getPhoneById(phoneId);
          if (phone) {
            phones.push(phone);
          }
        }
      }

      return phones;
    } catch (error) {
      console.error('Semantic search failed:', error);
      return this.fallbackProductSearch(query, filters, topK);
    }
  }

  /**
   * Fallback product search using database when Pinecone is unavailable
   */
  private async fallbackProductSearch(
    query: string,
    filters?: {
      maxPrice?: number;
      minPrice?: number;
      minRam?: number;
      minBattery?: number;
      company?: string;
    },
    limit: number = 5
  ): Promise<Phone[]> {
    const dbFilters: Filters = {};

    if (filters?.maxPrice) dbFilters.maxPrice = filters.maxPrice;
    if (filters?.minPrice) dbFilters.minPrice = filters.minPrice;
    if (filters?.minRam) dbFilters.minRam = filters.minRam;
    if (filters?.minBattery) dbFilters.minBattery = filters.minBattery;
    if (filters?.company) dbFilters.company = filters.company;

    const phones = await getFilteredPhones(dbFilters);
    return phones.slice(0, limit);
  }

  /**
   * Check if product index exists and has data
   */
  async hasProductIndex(): Promise<boolean> {
    if (!this.hasApiKey || !this.initialized) {
      return false;
    }

    try {
      const index = this.pinecone.index(this.indexName);
      const stats = await index.describeIndexStats();
      return (stats.totalRecordCount || 0) > 0;
    } catch {
      return false;
    }
  }

  /**
   * Delete and recreate the index with correct dimensions
   * Use this when switching embedding models (e.g., Gemini -> OpenAI)
   */
  async recreateIndex(): Promise<{ success: boolean; message: string }> {
    if (!this.hasApiKey) {
      return { success: false, message: 'Pinecone API key not configured' };
    }

    try {
      // Delete existing index if it exists
      const indexes = await this.pinecone.listIndexes();
      const indexExists = indexes.indexes?.some(idx => idx.name === this.indexName);

      if (indexExists) {
        await this.pinecone.deleteIndex(this.indexName);

        // Wait for deletion to complete
        await new Promise(resolve => setTimeout(resolve, 5000));
      }

      // Create new index with correct dimensions
      await this.pinecone.createIndex({
        name: this.indexName,
        dimension: AI_CONFIG.EMBEDDING_DIM,
        metric: 'cosine',
        spec: {
          serverless: {
            cloud: 'aws',
            region: 'us-east-1'
          }
        }
      });

      // Wait for index to be ready
      await this.waitForIndex();

      this.initialized = true;
      return { success: true, message: `Index recreated with dimension ${AI_CONFIG.EMBEDDING_DIM}` };
    } catch (error) {
      console.error('Failed to recreate index:', error);
      return { success: false, message: error instanceof Error ? error.message : 'Unknown error' };
    }
  }
}

// Singleton instance
let vectorClientInstance: VectorClient | null = null;

export function getVectorClient(): VectorClient {
  if (!vectorClientInstance) {
    vectorClientInstance = new VectorClient();
  }
  return vectorClientInstance;
}

File: backend/ai/vector_client.py
"""
Vector Client for Pinecone interactions.
Ported from src/lib/vector/vector-client.ts
"""

import os
from typing import Optional
from django.conf import settings

try:
    from pinecone import Pinecone
    PINECONE_AVAILABLE = True
except ImportError:
    PINECONE_AVAILABLE = False
    print("Pinecone not available, using fallback search")

from .llm_client import get_llm_client


class VectorClient:
    """
    Client for Pinecone vector database operations.
    Handles similarity search for typo correction and semantic product search.
    """
    
    _instance: Optional['VectorClient'] = None
    
    def __new__(cls):
        """Singleton pattern."""
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if self._initialized:
            return
        
        self.api_key = settings.PINECONE_API_KEY
        self.index_name = settings.PINECONE_INDEX_NAME
        self.pinecone = None
        self.index = None
        self.available = False
        
        if PINECONE_AVAILABLE and self.api_key:
            try:
                self.pinecone = Pinecone(api_key=self.api_key)
                self.index = self.pinecone.Index(self.index_name)
                self.available = True
            except Exception as e:
                print(f"Pinecone initialization error: {e}")
        
        self._initialized = True
    
    def find_similar(
        self,
        query: str,
        type_filter: str = None,
        company_filter: str = None,
        threshold: float = 0.7
    ) -> list:
        """
        Find similar company/model names for typo correction.
        
        Args:
            query: Search query
            type_filter: 'company' or 'model'
            company_filter: Filter by specific company
            threshold: Minimum similarity score
            
        Returns:
            List of similar matches with scores
        """
        if not self.available:
            return self._fallback_similar(query, type_filter)
        
        try:
            llm = get_llm_client()
            embedding = llm.embed(query)
            
            if not embedding:
                return self._fallback_similar(query, type_filter)
            
            # Build filter
            filter_dict = {}
            if type_filter:
                filter_dict['type'] = type_filter
            if company_filter:
                filter_dict['company'] = company_filter.lower()
            
            results = self.index.query(
                vector=embedding,
                top_k=5,
                include_metadata=True,
                filter=filter_dict if filter_dict else None
            )
            
            matches = []
            for match in results.get('matches', []):
                if match['score'] >= threshold:
                    matches.append({
                        'id': match['id'],
                        'score': match['score'],
                        'metadata': match.get('metadata', {})
                    })
            
            return matches
            
        except Exception as e:
            print(f"Vector search error: {e}")
            return self._fallback_similar(query, type_filter)
    
    def _fallback_similar(self, query: str, type_filter: str = None) -> list:
        """
        Fallback similarity search using string matching.
        """
        from api.models import Phone
        
        query_lower = query.lower()
        matches = []
        
        if type_filter == 'company' or type_filter is None:
            # Get unique companies
            companies = list(
                Phone.objects.values_list('company_name', flat=True).distinct()
            )
            for company in companies:
                score = self._string_similarity(query_lower, company.lower())
                if score >= 0.6:
                    matches.append({
                        'id': f"company_{company}",
                        'score': score,
                        'metadata': {'type': 'company', 'value': company}
                    })
        
        if type_filter == 'model' or type_filter is None:
            # Get models (limited)
            models = Phone.objects.values('model_name', 'company_name')[:100]
            for m in models:
                score = self._string_similarity(query_lower, m['model_name'].lower())
                if score >= 0.5:
                    matches.append({
                        'id': f"model_{m['model_name']}",
                        'score': score,
                        'metadata': {
                            'type': 'model',
                            'value': m['model_name'],
                            'company': m['company_name']
                        }
                    })
        
        # Sort by score descending
        matches.sort(key=lambda x: x['score'], reverse=True)
        return matches[:5]
    
    def _string_similarity(self, s1: str, s2: str) -> float:
        """
        Calculate string similarity using Levenshtein-like ratio.
        """
        if s1 == s2:
            return 1.0
        
        # Simple substring matching
        if s1 in s2 or s2 in s1:
            return 0.8
        
        # Character overlap
        set1 = set(s1)
        set2 = set(s2)
        intersection = len(set1 & set2)
        union = len(set1 | set2)
        
        return intersection / union if union > 0 else 0.0
    
    def search_products(
        self,
        query: str,
        filters: dict = None,
        top_k: int = 5
    ) -> list:
        """
        Semantic search for phones.
        
        Args:
            query: Natural language search query
            filters: Optional filters (company, price range, etc.)
            top_k: Number of results to return
            
        Returns:
            List of Phone objects
        """
        from api.models import Phone
        
        if not self.available:
            return self._fallback_search(query, filters, top_k)
        
        try:
            llm = get_llm_client()
            embedding = llm.embed(query)
            
            if not embedding:
                return self._fallback_search(query, filters, top_k)
            
            # Build Pinecone filter from filters dict
            pinecone_filter = {'type': 'product'}
            if filters:
                if filters.get('company'):
                    pinecone_filter['company'] = filters['company'].lower()
            
            results = self.index.query(
                vector=embedding,
                top_k=top_k,
                include_metadata=True,
                filter=pinecone_filter,
                namespace='products'
            )
            
            # Get phone IDs from results
            phone_ids = []
            for match in results.get('matches', []):
                phone_id = match.get('metadata', {}).get('phone_id')
                if phone_id:
                    phone_ids.append(int(phone_id))
            
            if not phone_ids:
                return self._fallback_search(query, filters, top_k)
            
            # Fetch full phone data
            phones = list(Phone.objects.filter(id__in=phone_ids))
            
            # Additional filtering
            if filters:
                phones = self._apply_filters(phones, filters)
            
            return phones
            
        except Exception as e:
            print(f"Product search error: {e}")
            return self._fallback_search(query, filters, top_k)
    
    def _fallback_search(self, query: str, filters: dict = None, top_k: int = 5) -> list:
        """
        Fallback search using database queries.
        """
        from api.models import Phone
        from django.db.models import Q
        
        phones = Phone.objects.all()
        query_lower = query.lower()
        
        # Keyword-based search
        keywords = query_lower.split()
        q_objects = Q()
        
        for keyword in keywords:
            q_objects |= Q(model_name__icontains=keyword)
            q_objects |= Q(company_name__icontains=keyword)
            q_objects |= Q(processor__icontains=keyword)
        
        if q_objects:
            phones = phones.filter(q_objects)
        
        # Apply filters
        if filters:
            phones = self._apply_filters_queryset(phones, filters)
        
        return list(phones.order_by('-user_rating')[:top_k])
    
    def _apply_filters(self, phones: list, filters: dict) -> list:
        """Apply filters to a list of Phone objects."""
        result = phones
        
        if filters.get('company'):
            result = [p for p in result if p.company_name.lower() == filters['company'].lower()]
        if filters.get('minPrice'):
            result = [p for p in result if p.price_inr >= int(filters['minPrice'])]
        if filters.get('maxPrice'):
            result = [p for p in result if p.price_inr <= int(filters['maxPrice'])]
        if filters.get('minRam'):
            result = [p for p in result if p.ram_gb >= float(filters['minRam'])]
        if filters.get('minBattery'):
            result = [p for p in result if p.battery_mah >= int(filters['minBattery'])]
        if filters.get('minCamera'):
            result = [p for p in result if p.back_camera_mp >= float(filters['minCamera'])]
        
        return result
    
    def _apply_filters_queryset(self, queryset, filters: dict):
        """Apply filters to a Django queryset."""
        if filters.get('company'):
            queryset = queryset.filter(company_name__icontains=filters['company'])
        if filters.get('minPrice'):
            queryset = queryset.filter(price_inr__gte=int(filters['minPrice']))
        if filters.get('maxPrice'):
            queryset = queryset.filter(price_inr__lte=int(filters['maxPrice']))
        if filters.get('minRam'):
            queryset = queryset.filter(ram_gb__gte=float(filters['minRam']))
        if filters.get('minBattery'):
            queryset = queryset.filter(battery_mah__gte=int(filters['minBattery']))
        if filters.get('minCamera'):
            queryset = queryset.filter(back_camera_mp__gte=float(filters['minCamera']))
        
        return queryset
    
    def build_product_index(self) -> dict:
        """
        Build/rebuild the product index in Pinecone.
        
        Returns:
            Status information
        """
        from api.models import Phone
        
        if not self.available:
            return {'success': False, 'error': 'Pinecone not available'}
        
        try:
            llm = get_llm_client()
            phones = Phone.objects.all()
            
            indexed = 0
            errors = 0
            
            for phone in phones:
                # Create rich description for embedding
                description = self._create_phone_description(phone)
                embedding = llm.embed(description)
                
                if not embedding:
                    errors += 1
                    continue
                
                # Upsert to Pinecone
                self.index.upsert(
                    vectors=[{
                        'id': f"phone_{phone.id}",
                        'values': embedding,
                        'metadata': {
                            'type': 'product',
                            'phone_id': phone.id,
                            'company': phone.company_name.lower(),
                            'model': phone.model_name.lower(),
                            'price': phone.price_inr,
                        }
                    }],
                    namespace='products'
                )
                indexed += 1
            
            return {
                'success': True,
                'indexed': indexed,
                'errors': errors,
                'total': phones.count()
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def _create_phone_description(self, phone) -> str:
        """Create a rich text description for embedding."""
        features = []
        
        # Price tier
        if phone.price_inr < 15000:
            features.append("budget affordable cheap economical")
        elif phone.price_inr < 30000:
            features.append("mid-range balanced value")
        elif phone.price_inr < 50000:
            features.append("upper mid-range premium features")
        else:
            features.append("flagship premium high-end")
        
        # Camera
        if phone.back_camera_mp >= 100:
            features.append("excellent camera photography flagship camera")
        elif phone.back_camera_mp >= 50:
            features.append("good camera photography")
        
        # Battery
        if phone.battery_mah >= 5000:
            features.append("long battery life all-day battery")
        
        # RAM (gaming)
        if phone.ram_gb >= 8:
            features.append("gaming smooth performance multitasking")
        
        return f"""
{phone.company_name} {phone.model_name} {phone.memory_gb}gb.
Price ₹{phone.price_inr:,}. {phone.ram_gb}GB RAM, {phone.memory_gb}GB storage.
{phone.back_camera_mp}MP rear camera, {phone.front_camera_mp}MP front camera.
{phone.battery_mah}mAh battery. {phone.screen_size}" display.
Processor: {phone.processor or 'Unknown'}.
Rating: {phone.user_rating}/5.
Features: {' '.join(features)}.
""".strip()


# Singleton getter
_vector_client = None

def get_vector_client() -> VectorClient:
    """Get or create the Vector client singleton."""
    global _vector_client
    if _vector_client is None:
        _vector_client = VectorClient()
    return _vector_client
